homeassistant/components/tplink/entity.py

```diff
@@ -62,24 +121,294 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+        # feature-based entities use feature ids to create unique ids.
+        # When a new non-feature based platform is added,
+        # it needs to implement this.
+        raise NotImplementedError
+
+    async def async_added_to_hass(self) -> None:
+        """Handle being added to hass."""
+        self._async_call_update_attrs()
+        return await super().async_added_to_hass()
+
+    @abstractmethod
+    @callback
+    def _async_update_attrs(self) -> None:
+        """Platforms implement this to update the entity internals."""
+        raise NotImplementedError
+
+    @callback
+    def _async_call_update_attrs(self) -> None:
+        """Call update_attrs and make entity unavailable on error.
+
+        update_attrs can sometimes fail if a device firmware update breaks the
+        downstream library.
+        """
+        try:
+            self._async_update_attrs()
+        except Exception as ex:  # noqa: BLE001
+            if self._attr_available:
+                _LOGGER.warning(
+                    "Unable to read data for %s %s: %s",
+                    self._device,
+                    self.entity_id,
+                    ex,
+                )
+            self._attr_available = False
+        else:
+            self._attr_available = True
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        self._async_call_update_attrs()
+        super()._handle_coordinator_update()
+
+    @property
+    def available(self) -> bool:
+        """Return if entity is available."""
+        return self.coordinator.last_update_success and self._attr_available
+
+
+class CoordinatedTPLinkFeatureEntity(CoordinatedTPLinkEntity, ABC):
+    """Common base class for all coordinated tplink feature entities."""
+
+    entity_description: TPLinkFeatureEntityDescription
+    _feature: Feature
+
+    def __init__(
+        self,
+        device: Device,
+        coordinator: TPLinkDataUpdateCoordinator,
+        *,
+        feature: Feature,
+        description: TPLinkFeatureEntityDescription,
+        parent: Device | None = None,
+    ) -> None:
+        """Initialize the entity."""
+        self.entity_description = description
+        super().__init__(device, coordinator, parent=parent, feature=feature)
+
+    def _get_unique_id(self) -> str:
+        """Return unique ID for the entity."""
+        key = self.entity_description.key
+        # The unique id for the state feature in the switch platform is the device_id
+        if key == PRIMARY_STATE_ID:
+            return legacy_device_id(self._device)
+
+        # Historically the legacy device emeter attributes which are now replaced with
+        # features used slightly different keys. This ensures that those entities
+        # are not orphaned. Returns the mapped key or the provided key if not mapped.
+        key = LEGACY_KEY_MAPPING.get(key, key)
```

Comment:
Please split long lines at 80

Reply:
Done in https://github.com/home-assistant/core/pull/120404